File: typescript-sdk/src/http.ts
import {
  AuthenticationError,
  ConflictError,
  NotFoundError,
  RateLimitError,
  ServerError,
  TimeoutError,
  ValidationError
} from "./errors";

const sleep = (ms: number) => new Promise((resolve) => setTimeout(resolve, ms));

export interface HttpRequestOptions extends RequestInit {
  timeoutMs?: number;
  retryAttempts?: number;
}

export class HttpClient {
  constructor(
    private readonly baseUrl: string,
    private readonly apiKey: string,
    private readonly defaultTimeoutMs = 120_000,
    private readonly defaultRetryAttempts = 3
  ) {}

  async request<T>(path: string, options: HttpRequestOptions = {}): Promise<T> {
    const timeoutMs = options.timeoutMs ?? this.defaultTimeoutMs;
    const retryAttempts = options.retryAttempts ?? this.defaultRetryAttempts;
    const url = `${this.baseUrl}${path}`;
    const headers: HeadersInit = {
      Authorization: `Bearer ${this.apiKey}`,
      "Content-Type": "application/json",
      ...(options.headers ?? {})
    };

    let lastError: unknown;

    for (let attempt = 0; attempt <= retryAttempts; attempt += 1) {
      const controller = new AbortController();
      const timeout = setTimeout(() => controller.abort(), timeoutMs);

      try {
        const response = await fetch(url, { ...options, headers, signal: controller.signal });
        clearTimeout(timeout);

        if (response.ok) {
          const contentType = response.headers.get("content-type") ?? "";
          if (contentType.includes("application/json")) {
            return (await response.json()) as T;
          }
          return (response.body as T) ?? ((await response.text()) as T);
        }

        const responseText = await response.text();
        const retryable = response.status === 408 || response.status === 429 || response.status >= 500;
        if (retryable && attempt < retryAttempts) {
          const delay = Math.min(300 * 2 ** attempt + Math.floor(Math.random() * 100), 10_000);
          await sleep(delay);
          continue;
        }

        if (response.status === 401 || response.status === 403) throw new AuthenticationError(responseText);
        if (response.status === 404) throw new NotFoundError(responseText);
        if (response.status === 409) throw new ConflictError(responseText);
        if (response.status === 429) throw new RateLimitError(responseText);
        if (response.status >= 400 && response.status < 500) throw new ValidationError(responseText);
        throw new ServerError(responseText);
      } catch (error) {
        clearTimeout(timeout);
        if (error instanceof DOMException && error.name === "AbortError") {
          lastError = new TimeoutError(`Request timed out after ${timeoutMs}ms`);
        } else {
          lastError = error;
        }
        if (attempt < retryAttempts) {
          const delay = Math.min(300 * 2 ** attempt + Math.floor(Math.random() * 100), 10_000);
          await sleep(delay);
          continue;
        }
      }
    }

    throw lastError;
  }
}

File: python-sdk/src/dispersl/http.py
from __future__ import annotations

import asyncio
import random
from typing import Any

import httpx

from .errors import (
    AuthenticationError,
    ConflictError,
    NotFoundError,
    RateLimitError,
    ServerError,
    TimeoutError,
    ValidationError,
)


def _map_status(code: int, message: str) -> Exception:
    if code in {401, 403}:
        return AuthenticationError(message)
    if code == 404:
        return NotFoundError(message)
    if code == 409:
        return ConflictError(message)
    if code == 429:
        return RateLimitError(message)
    if 400 <= code < 500:
        return ValidationError(message)
    return ServerError(message)


class AsyncHttpClient:
    def __init__(self, base_url: str, api_key: str, timeout_s: float = 120.0, retry_attempts: int = 3) -> None:
        self.base_url = base_url.rstrip("/")
        self.timeout_s = timeout_s
        self.retry_attempts = retry_attempts
        self.client = httpx.AsyncClient(
            base_url=self.base_url,
            timeout=self.timeout_s,
            headers={"Authorization": f"Bearer {api_key}", "Content-Type": "application/json"},
        )

    async def request(self, method: str, path: str, json_body: dict[str, Any] | None = None) -> Any:
        last_error: Exception | None = None
        for attempt in range(self.retry_attempts + 1):
            try:
                response = await self.client.request(method, path, json=json_body)
                if response.status_code < 400:
                    if "application/json" in response.headers.get("content-type", ""):
                        return response.json()
                    return response

                retryable = response.status_code in {408, 429} or response.status_code >= 500
                if retryable and attempt < self.retry_attempts:
                    delay = min(0.3 * (2**attempt) + random.random() * 0.1, 10.0)
                    await asyncio.sleep(delay)
                    continue
                raise _map_status(response.status_code, response.text)
            except httpx.TimeoutException as exc:
                last_error = TimeoutError(str(exc))
                if attempt < self.retry_attempts:
                    delay = min(0.3 * (2**attempt) + random.random() * 0.1, 10.0)
                    await asyncio.sleep(delay)
                    continue
            except Exception as exc:
                last_error = exc if isinstance(exc, Exception) else Exception(str(exc))
                if attempt < self.retry_attempts:
                    delay = min(0.3 * (2**attempt) + random.random() * 0.1, 10.0)
                    await asyncio.sleep(delay)
                    continue
        raise last_error if last_error else RuntimeError("Request failed")

    async def aclose(self) -> None:
        await self.client.aclose()
